File: 02/ICPC2021-Prelim/B/ans1.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i, n) for (int i = 0; i < (n); ++i)

int main() {
  // まだ確定していないときの値（10^9とした）
  constexpr int Undefined = 1e9;

  while (true) {
    // 入力
    int w, h;
    cin >> w >> h;

    // もしwとhが両方0であれば、終了
    if (w + h == 0) break;

    // 決められたマスを受け取る
    vector Grid(w, vector<int>(h, Undefined));
    rep(_, w + h - 1) {
      int x, y, n;
      cin >> x >> y >> n;
      Grid[x - 1][y - 1] = n;
    }

    // 上端・左端のマスの配列
    vector<int> Top(w, Undefined), Left(h, Undefined);
    // 上端の一番左を0にする
    Top[0] = 0;

    while (true) {
      // 更新されたかを管理する
      bool updated = false;

      rep(i, w) rep(j, h) {
        // 上・左が決まっていて、マスが未確定なら決める
        if (Top[i] != Undefined && Left[j] != Undefined && Grid[i][j] == Undefined) {
          Grid[i][j] = Top[i] + Left[j];
          updated = true;
        }

        // 左とマスが決まっているが、上が決まっていない
        if (Left[j] != Undefined && Grid[i][j] != Undefined && Top[i] == Undefined) {
          Top[i] = Grid[i][j] - Left[j];
          updated = true;
        }

        // 上とマスが決まっているが、左が決まっていない
        if (Top[i] != Undefined && Grid[i][j] != Undefined && Left[j] == Undefined) {
          Left[j] = Grid[i][j] - Top[i];
          updated = true;
        }
      }

      // 更新されなくなったら、終了
      if (!updated) break;
    }

    // 未確定の場所があるか調べる
    bool existsUndefined = false;
    rep(i, w) if (Top[i] == Undefined) existsUndefined = true;
    rep(i, h) if (Left[i] == Undefined) existsUndefined = true;
    rep(i, w) rep(j, h) {
      if (Grid[i][j] == Undefined) existsUndefined = true;
    }

    // 未確定の場所があればNo、そうでなければYes。
    cout << (existsUndefined ? "NO" : "YES") << endl;
  }
  return 0;
}

File: 02/ICPC2021-Prelim/B/ans1.py
# まだ確定していないときの値（10^9）とした
UNDEFINED = 10**9

while True:
    # 入力
    w, h = map(int, input().split())

    # もしwとhが両方0であれば、終了
    if w + h == 0:
        break

    # 決められたマスを受け取る
    Grid = [[UNDEFINED] * h for _ in range(w)]
    for _ in range(w+h-1):
        x, y, n = map(int, input().split())
        Grid[x-1][y-1] = n

    # 上端・左端のマスの配列
    Top = [UNDEFINED] * w
    Left = [UNDEFINED] * h

    # 上端の一番左を0にする
    Top[0] = 0

    while True:
        # 更新されたかを管理する
        updated = False

        for i in range(w):
            for j in range(h):
                # 上・左が決まっていて、マスが未確定なら決める
                if Top[i] != UNDEFINED and Left[j] != UNDEFINED and Grid[i][j] == UNDEFINED:
                    Grid[i][j] = Top[i] + Left[j]
                    updated = True

                # 左とマスが決まっているが、上が決まっていない
                if Left[j] != UNDEFINED and Grid[i][j] != UNDEFINED and Top[i] == UNDEFINED:
                    Top[i] = Grid[i][j] - Left[j]
                    updated = True

                # 上とマスが決まっているが、左が決まっていない
                if Top[i] != UNDEFINED and Grid[i][j] != UNDEFINED and Left[j] == UNDEFINED:
                    Left[j] = Grid[i][j] - Top[i]
                    updated = True

        # 更新されなかったら終了
        if not updated:
            break

    # 未確定の場所があるか調べる
    existsUndefined = False
    for i in range(w):
        if Top[i] == UNDEFINED:
            existsUndefined = True
    for i in range(h):
        if Left[i] == UNDEFINED:
            existsUndefined = True
    for i in range(w):
        for j in range(h):
            if Grid[i][j] == UNDEFINED:
                existsUndefined = True

    # 未確定の場所があればNo、そうでなければYes。
    print("NO" if existsUndefined else "YES")
